### PR TITLE
ipq806x: improve support for Askey RT4230W REV6

### PR DESCRIPTION
**Pull request by aiamadeus**
- Fix caldata script
The partition name in the device dts is '0:ART'.
Be independent to prevent this part from becoming
incorrect once the kernel v5.4 gone.
Fixes: da8428d ("ipq806x: add support for Askey RT4230W REV6")
- Use usual writing for pcie part
The problem has been fixed in f47cb40 ("ipq806x: fix pci broken on bootm
command"), now the pcie part can be written in the usual way.
- Enable onboard spi flash
There is a mr25h256 spi flash on this machine. From the mtd backup
of the original firmware, this spi flash is empty.

### CODE REVIEW

**Comment by adschm** [comment]
@pkgadd @Ansuel Could you review/comment?

**Comment by Ansuel** [comment]
@AmadeusGhost didn't check the dts but this does use the smem parser of the partition are statically defined?

**Comment by AmadeusGhost** [comment]
> @AmadeusGhost didn't check the dts but this does use the smem parser of the partition are statically defined?

It seems that neither uboot (smeminfo) nor the stock firmware defines any partitions on the spi flash.
There are only a few lines of related log in the stock firmware:
```
[    2.933211] spi_qup 1a280000.spi: IN:block:16, fifo:64, OUT:block:16, fifo:64
[    2.939489] spi_qup 1a280000.spi: Failed to allocate dma rx chan
[    2.945404] spi_qup 1a280000.spi: Failed to allocate dma tx chan
[    2.952788] m25p80 spi32766.0: found mr25h256, expected s25fl256s1
[    2.958040] m25p80 spi32766.0: mr25h256 (32 Kbytes)
```

**Comment by Ansuel** [comment]
i mean for the calibration data

**Comment by AmadeusGhost** [comment]
> i mean for the calibration data

I'm sorry I don't know what you mean very much.
If you are referring to the partition name definition of art, it's only defined in the dts of this device, uboot smeminfo does not have it.

**Comment by Ansuel** [comment]
I was concerned about that... in that case yes you are right
@adschm LGTM

**Comment by pkgadd** [comment]
Without owning the device in question, nor considering myself to be a DTS specialist, I do think these patches are fine. At least "ipq806x: Askey RT4230W REV6: fix caldata script" is indeed necessary, while it works right now with the upper case fallback in place, that fallback will be dropped with kernel v5.4 in the near future - so having this fixed and explicit is (very) good (the device hardcodes the partitions via DTS as upper-case and doesn't use qcom-smem).

@adschm LGTM (to me as well)

File: target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts [hunk 1]
```diff
@@ -96,6 +96,35 @@
...
+
+	spi4: spi@1a280000 {
+		status = "okay";
+
+		pinctrl-0 = <&spi_pins>;
+		pinctrl-names = "default";
+
+		cs-gpios = <&qcom_pinmux 20 GPIO_ACTIVE_HIGH>;
+
+		m25p80@0 {
```
Comment (adschm) [inline target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:120]: flash@0

File: target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts [hunk 2]
```diff
@@ -96,6 +96,35 @@
...
+			drive-strength = <12>;
+		};
+	};
+};
+
+&gsbi5 {
+	qcom,mode = <GSBI_PROT_SPI>;
+	status = "okay";
+
+	spi4: spi@1a280000 {
```
Comment (adschm) [inline target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:112]: I don't think we need the spi4 label here.

Comment (AmadeusGhost) [reply target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts]: > I don't think we need the spi4 label here.

Ok, get it.

**Comment by Ansuel** [comment]
@adschm worth a new review?

**Comment by adschm** [comment]
I've picked the first commit for now, for the other ones I'd like some explicit reviewed-by ...

**Comment by Ansuel** [comment]
Will check them ASAP

Il Mar 24 Ago 2021, 18:21 Adrian Schmutzler ***@***.***> ha
scritto:

**Comment by Ansuel** [comment]
Reviewed-by:  Ansuel Smith ansuelsmth@gmail.com

@adschm 
(dubious about the spi but since it's empty... no problem...)
(the pcie commit is important since we changed the pci implementation and force_gen1 doesn't exist anymore)

File: target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts [hunk 3]
```diff
@@ -96,6 +96,35 @@
...
+	spi@1a280000 {
+		status = "okay";
+
+		pinctrl-0 = <&spi_pins>;
+		pinctrl-names = "default";
+
+		cs-gpios = <&qcom_pinmux 20 GPIO_ACTIVE_HIGH>;
+
+		flash@0 {
+			compatible = "everspin,mr25h256";
```
Comment (adschm) [inline target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:121]: jedec,spi-nor does not work?

Comment (AmadeusGhost) [reply target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:121]: > jedec,spi-nor does not work?

Yes, "jedec, spi-nor" does not work. See also: https://github.com/torvalds/linux/commit/5ca11ca7ad950c60d4a30bbf9eb18555d537b226

**Comment by AmadeusGhost** [comment]
> Reviewed-by: Ansuel Smith [ansuelsmth@gmail.com](mailto:ansuelsmth@gmail.com)

Thank you very much.

**Comment by Ansuel** [comment]
@AmadeusGhost can you convert this to nvmem api ? https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=20470580cd271de1e6017c88261b577ed4773476

Aside from that LGTM...

**Comment by AmadeusGhost** [comment]
> @AmadeusGhost can you convert this to nvmem api ?

Done, everything is working fine except this:
(Add this `qcom,boot_pages_size=<0x1180000>;` didn't work)
```
[   57.742362] print_req_error: 174 callbacks suppressed
[   57.742373] blk_update_request: I/O error, dev mtdblock0, sector 0 op 0x0:(READ) flags 0x80700 phys_seg 4 prio class 0
[   57.747313] blk_update_request: I/O error, dev mtdblock0, sector 8 op 0x0:(READ) flags 0x80700 phys_seg 3 prio class 0
[   57.757967] blk_update_request: I/O error, dev mtdblock0, sector 16 op 0x0:(READ) flags 0x80700 phys_seg 2 prio class 0
[   57.768794] blk_update_request: I/O error, dev mtdblock0, sector 24 op 0x0:(READ) flags 0x80700 phys_seg 1 prio class 0
[   57.779562] blk_update_request: I/O error, dev mtdblock0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[   57.789116] buffer_io_error: 46 callbacks suppressed
[   57.789121] Buffer I/O error on dev mtdblock0, logical block 0, async page read
[   57.829837] blk_update_request: I/O error, dev mtdblock0, sector 8 op 0x0:(READ) flags 0x80700 phys_seg 1 prio class 0
[   57.839796] blk_update_request: I/O error, dev mtdblock0, sector 8 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[   57.839824] Buffer I/O error on dev mtdblock0, logical block 1, async page read
[   57.850863] blk_update_request: I/O error, dev mtdblock1, sector 0 op 0x0:(READ) flags 0x80700 phys_seg 4 prio class 0
[   57.857836] blk_update_request: I/O error, dev mtdblock1, sector 8 op 0x0:(READ) flags 0x80700 phys_seg 3 prio class 0
[   57.868708] blk_update_request: I/O error, dev mtdblock1, sector 16 op 0x0:(READ) flags 0x80700 phys_seg 2 prio class 0
[   57.969677] Buffer I/O error on dev mtdblock1, logical block 0, async page read
[   57.974120] Buffer I/O error on dev mtdblock1, logical block 1, async page read
[   58.045272] Buffer I/O error on dev mtdblock12, logical block 0, async page read
[   58.048702] Buffer I/O error on dev mtdblock12, logical block 1, async page read
[   58.059217] Buffer I/O error on dev mtdblock13, logical block 0, async page read
[   58.065071] Buffer I/O error on dev mtdblock13, logical block 0, async page read
[   58.072440] Buffer I/O error on dev mtdblock14, logical block 0, async page read
[   58.077934] Buffer I/O error on dev mtdblock14, logical block 0, async page read
```

**Comment by mans0n** [comment]
The commit title is a bit long (especially `Askey RT4230W REV6:`), could you reduce it somehow?
And please add Ansuel's Reviewed-by to the commit message.

> Done, everything is working fine except this:

Is this error from nvmem?

**Comment by AmadeusGhost** [comment]
> And please add Ansuel's Reviewed-by to the commit message.

@Ansuel I would appreciate it if you could review [this](1887d080e6bac0dc23a155c918d1ffacc2a82371).
> Is this error from nvmem?

No, this "Buffer I/O error" error existed before.

**Comment by Ansuel** [comment]
@AmadeusGhost can you test `qcom,boot_pages_size=<0x2400000>;` after 
`nand-ecc-step-size = <512>;` also you need `nand-is-boot-medium = "true";`

**Comment by AmadeusGhost** [comment]
> @AmadeusGhost can you test `qcom,boot_pages_size=<0x2400000>;` after `nand-ecc-step-size = <512>;` also you need `nand-is-boot-medium = "true";`

Thanks for your reply, `qcom,boot_pages_size=<0x2400000>;` will cause the mac address of the wired network port to be random (It seems that gmac failed to read nvmem-cells).

**Comment by hauke** [comment]
Thank you for your patches, I applied them to master in eee41e33eca2f860724bceda3f36ea2e30149ef0 and 45b3111b992377f0298e52e8787b1a730440b3bd.

**Comment by Ansuel** [comment]
@AmadeusGhost Ehy can you test one patch? I reworked the patch for sparse partition. Wonder if you want to test it.

**Comment by Lanchon** [comment]
@Ansuel i wanted to patch this device's dts. where is your patch?

**Comment by AmadeusGhost** [comment]
> Ehy can you test one patch? I reworked the patch for sparse partition. Wonder if you want to test it.

Of course I can, but I may not be able to reply in time due to busy study.

**Pull request closed without merging**